go-controller/pkg/cluster/ha_master.go

```diff
@@ -0,0 +1,321 @@
+			}
+
+			// Get the master ip
+			masterIPList, sbDBPort, nbDBPort, err := extractDbRemotesFromEndpoint(ep)
+			if err != nil {
+				return false, nil
+			}
+
+			if hacluster.currentLeaderIP == masterIPList[0] {
+				// Looks like new master has not yet updated the ovnkube-db endpoint.
+				// Wait for 500 ms.
+				return false, nil
+			}
+
+			hacluster.currentLeaderIP = masterIPList[0]
+			logrus.Infof("New leader IP is : [%s]", hacluster.currentLeaderIP)
+
+			err = hacluster.DemoteOVNDbs(hacluster.currentLeaderIP, nbDBPort, sbDBPort)
+			if err != nil {
+				logrus.Errorf("Demoting OVN ovsdb-servers to standby failed")
+				return false, err
+			}
+
+			return true, nil
+		}); err != nil {
+			logrus.Errorf("Error getting the correct endpoint [%s]", ovnkubeDbEp)
+			return err
+		}
+	}
+
+	hacluster.cluster.StopClusterMaster()
+
+	return nil
+}
+
+//PromoteOVNDbs promotes the OVN Db servers and resumes the ovn-northd.
+func (hacluster *OvnHAMasterController) PromoteOVNDbs(nbDBPort, sbDBPort string) error {
+	_, _, err := util.RunOVNCtl("promote_ovnnb")
+	if err != nil {
+		logrus.Errorf("promoting ovnnb failed")
+		return err
+	}
+	_, _, err = util.RunOVNCtl("promote_ovnsb")
+	if err != nil {
+		logrus.Errorf("promoting ovnsb failed")
+		return err
+	}
+
+	// Configure OVN dbs to listen on the ovnkube-pod-ip
+	target := "ptcp:" + nbDBPort + ":" + config.Kubernetes.OVNKubePodIP
+	_, _, err = util.RunOVNNBAppCtl("ovsdb-server/add-remote", target)
+	if err != nil {
+		logrus.Errorf("Adding remote [%s] to NB ovsdb-server failed", target)
+		return err
+	}
+
+	target = "ptcp:" + sbDBPort + ":" + config.Kubernetes.OVNKubePodIP
+	_, _, err = util.RunOVNSBAppCtl("ovsdb-server/add-remote", target)
+	if err != nil {
+		logrus.Errorf("Adding remote [%s] to SB ovsdb-server failed", target)
+		return err
+	}
+
+	// Ignore the error from this command for now as the patch in OVN
+	// to add the pause/resume support is not yet merged.
+	//TODO: Handle the err properly when ovn-northd supports pause/resume
+	_, _, _ = util.RunOVNNorthAppCtl("resume")
+	return nil
+}
+
+//DemoteOVNDbs demotes the OVN Db servers and configure them to connect
+//to the new master and pauses the ovn-northd.
+func (hacluster *OvnHAMasterController) DemoteOVNDbs(masterIP, nbDBPort, sbDBPort string) error {
+	_, _, err := util.RunOVNCtl("demote_ovnnb", "--db-nb-sync-from-addr="+masterIP,
+		"--db-nb-sync-from-port="+nbDBPort)
+	if err != nil {
+		logrus.Errorf("Demoting NB ovsdb-server failed")
+		return err
+	}
+
+	_, _, err = util.RunOVNCtl("demote_ovnsb", "--db-sb-sync-from-addr="+masterIP,
+		"--db-sb-sync-from-port="+sbDBPort)
+
+	if err != nil {
+		logrus.Errorf("Demoting SB ovsdb-server failed")
+		return err
+	}
+
+	// Ignore the error from this command for now as the patch in OVN
+	// to add the pause/resume support is not yet merged.
+	//TODO: Handle the err properly when ovn-northd supports pause/resume
+	_, _, _ = util.RunOVNNorthAppCtl("pause")
```

Comment:
Does it make sense to stop ovn-northd for now?

Reply:
What happens if we stop it ?

Does k8s  restart ovn-northd cotainer ?

Reply:
Oh duh, it totally will.

Also, is "pausing" permanent? If ovn-northd is randomly restarted, will it re-start in a paused state? Or will it re-resume processing?

Reply:
When it restarts it will be running fresh. So it will be resuming processing.
All the ovs-appctl commands to ovn-northd, ovn-controller, ovsdb-servers (and even ovs-vswitchd) are short lived and are not stored in the db.